Sources/NIOCore/ByteBuffer-QUICLengthPrefix.swift

```diff
@@ -0,0 +1,238 @@
+        var written = 0
+        written += self.writeQUICVariableLengthInteger(buffer.readableBytes)
+        written += self.writeImmutableBuffer(buffer)
+        return written
+    }
+
+    /// Write a QUIC variable-length integer prefixed string.
+    ///
+    /// [RFC 9000 § 16](https://www.rfc-editor.org/rfc/rfc9000.html#section-16)
+    ///
+    /// Write `string` prefixed with the length of string as a QUIC variable-length integer
+    /// - Parameter string: The string to be written
+    /// - Returns: The total bytes written. This is the bytes needed to write the length, plus the length of the string itself
+    @discardableResult
+    public mutating func writeQUICVariableLengthPrefixedString(_ string: String) -> Int {
+        var written = 0
+        // writeString always writes the String as UTF8 bytes, without a null-terminator
+        // So the length will be the utf8 count
+        written += self.writeQUICVariableLengthInteger(string.utf8.count)
+        written += self.writeString(string)
+        return written
+    }
+
+    /// Write a QUIC variable-length integer prefixed collection of bytes.
+    ///
+    /// [RFC 9000 § 16](https://www.rfc-editor.org/rfc/rfc9000.html#section-16)
+    ///
+    /// Write `bytes` prefixed with the number of bytes as a QUIC variable-length integer
+    /// - Parameter bytes: The bytes to be written
+    /// - Returns: The total bytes written. This is the bytes needed to write the length, plus the length of the string itself
+    @inlinable
+    public mutating func writeQUICVariableLengthPrefixedBytes<Bytes: Sequence>(_ bytes: Bytes) -> Int
+    where Bytes.Element == UInt8 {
+        let numberOfBytes = bytes.withContiguousStorageIfAvailable { b in
+            UnsafeRawBufferPointer(b).count
+        }
+        if let numberOfBytes {
+            var written = 0
+            written += self.writeQUICVariableLengthInteger(numberOfBytes)
+            written += self.writeBytes(bytes)
+            return written
+        } else {
+            return self.writeQUICVariableLengthPrefixed { buffer in
+                buffer.writeBytes(bytes)
+            }
+        }
+    }
+}
+
+extension ByteBuffer {
+    /// Creates `requiredSpace` bytes of free space before `index` by moving the `length` bytes after `index` forward by `requiredSpace`
+    /// e.g. given [a, b, c, d, e, f, g, h, i, j] and calling this function with (before: 4, length: 6, requiredSpace: 2) would result in
+    /// [a, b, c, d, 0, 0, e, f, g, h, i, j]
+    /// 2 extra bytes of space were created before index 4 (the letter e)
+    /// The total bytes written will be equal to `requiredSpace`, and the writer index will be moved accordingly
+    @usableFromInline
+    mutating func _createSpace(before index: Int, length: Int, requiredSpace: Int) {
+        // Add the required number of bytes to the end first
+        self.writeBytes([UInt8](repeatElement(0, count: requiredSpace)))
```

Comment:
Better to force this to 4 and write the integer than to make the allocation here.